test/jb_splits_store/set.test.js

```diff
@@ -0,0 +1,278 @@
+  let SET_SPLITS_PERMISSION_INDEX;
+
+  before(async function () {
+    let jbOperationsFactory = await ethers.getContractFactory('JBOperations');
+    let jbOperations = await jbOperationsFactory.deploy();
+    SET_SPLITS_PERMISSION_INDEX = await jbOperations.SET_SPLITS();
+  });
+
+  async function setup() {
+    let [deployer, projectOwner, ...addrs] = await ethers.getSigners();
+
+    let mockJbOperatorStore = await deployMockContract(deployer, jbOperatorStore.abi);
+    let mockJbProjects = await deployMockContract(deployer, jbProjects.abi);
+    let mockJbDirectory = await deployMockContract(deployer, jbDirectory.abi);
+
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(projectOwner.address, projectOwner.address, PROJECT_ID, SET_SPLITS_PERMISSION_INDEX)
+      .returns(true);
+
+    await mockJbProjects.mock.ownerOf
+      .withArgs(PROJECT_ID)
+      .returns(projectOwner.address);
+
+    await mockJbDirectory.mock.controllerOf
+      .withArgs(PROJECT_ID)
+      .returns(projectOwner.address);
+
+    let jbSplitsStoreFact = await ethers.getContractFactory('JBSplitsStore');
+    let jbSplitsStore = await jbSplitsStoreFact.deploy(
+      mockJbOperatorStore.address,
+      mockJbProjects.address,
+      mockJbDirectory.address
+    );
+
+    let splits = makeSplits(addrs[0].address);
+
+    return { deployer, projectOwner, addrs, jbSplitsStore, splits, mockJbOperatorStore, mockJbProjects, mockJbDirectory };
+  }
+
+  function makeSplits(beneficiaryAddress, count=4) {
+    let splits = []
+    for (let i = 0; i < count; i++) {
+      splits.push({
+        preferClaimed: false,
+        percent: Math.floor(10000000 / count),
+        lockedUntil: 0,
+        beneficiary: beneficiaryAddress,
+        allocator: ethers.constants.AddressZero,
+        projectId: 0
+      });
+    }
+    return splits;
+  }
+
+  it('set(...) and corresponding events', async function () {
+    const { projectOwner, addrs, jbSplitsStore, splits, mockJbOperatorStore, mockJbDirectory } = await setup();
+
+    await mockJbOperatorStore.mock.hasPermission.returns(false);
+    await mockJbDirectory.mock.controllerOf.returns(addrs[0].address);
+
+    const tx = await jbSplitsStore.connect(projectOwner).set(
+      PROJECT_ID,
+      DOMAIN,
+      GROUP,
+      splits
+    );
+
+    // Expect one event per split in splits[]
+    await Promise.all(
+      splits.map(async (split, _) => {
+        await expect(tx)
+          .to.emit(jbSplitsStore, 'SetSplit')
+          .withArgs(PROJECT_ID, DOMAIN, GROUP, Object.values(split), projectOwner.address);
+      }),
+    );
+
+    // Get the current splits (for this proj/dom/group)
+    let splitsStored = await jbSplitsStore.splitsOf(PROJECT_ID, DOMAIN, GROUP);
+
+    //compare every currently stored splits to the one we've just sent
+    for (let [idx, split] of splitsStored) {
+      for (let splitKey of Object.keys(split)) {
+        expect(split.splitKey).to.equal(splits[idx].splitKey);
+      }
```

Comment:
Why do we need to iterate over all of the keys? why can't we just compare the full object? i think chai has an expectation for this: https://medium.com/building-ibotta/testing-arrays-and-objects-with-chai-js-4b372310fe6d

expect(split).to.eql(splits[idx])

Reply:
If this works replace other instances.

Reply:
Actually, can we not just directly compare splitsStored and splits?

Reply:
Issue is how ethers is returning the value, as a Result object (which is extending Array). As soon as I compare it in Chai (even for deep equality as you suggest, or just via include.ordered), it behaves as an Array (and I therefore "loose" the keys, resulting in assertion failed).

Agree, there should be another (easier) way, will try again later today

This is as "expected/actual" (for both eql on whole objects or iterating):
      -  [
      -    false
      -    2500000
      -    0
      -    "0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"
      -    "0x0000000000000000000000000000000000000000"
      -    {
      -      "_hex": "0x00"
      -      "_isBigNumber": true
      -    }
      -  ]
      +  {
      +    "allocator": "0x0000000000000000000000000000000000000000"
      +    "beneficiary": "0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"
      +    "lockedUntil": 0
      +    "percent": 2500000
      +    "preferClaimed": false
      +    "projectId": 0
      +  }

Reply:
As we discussed over chat. Let's make sure this is fixed, as the current logic is not verifying the splits